File: Striver 79 Last Moment DSA Sheet/Day 1 - Arrays and Hashing/maximum-product-subarray.cpp

// https://leetcode.com/problems/maximum-product-subarray/description/

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ans = INT_MIN, p1 = 1, p2 = 1, size = nums.size();

        for(int i = 0; i < size; i++){
            int j = size-i-1;
            p1 = nums[i]*p1;
            p2 = nums[j]*p2;
            ans = max({ans, p1, p2});
            if(p1 == 0) p1 = 1;
            if(p2 == 0) p2 = 1;
        }

        return ans;
    }
};

File: Striver Important Interview Questions/Day 25 - Dynamic Programming/maximum-product-subarray.js

// https://leetcode.com/problems/maximum-product-subarray/

/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    const n = nums.length;
    let ans = Number.MIN_SAFE_INTEGER;
    if(n === 0) return ans;

    let p1 = 1, p2 = 1;
    for(let i = 0; i < n; i++){
        let j = n-i-1;
        p1 = nums[i]*p1;
        p2 = nums[j]*p2;
        ans = Math.max(ans, p1, p2);
        // console.log(ans);
        if(p1 === 0) p1 = 1;
        if(p2 === 0) p2 = 1;
    }

    return ans;
};